presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java

```diff
@@ -139,6 +139,16 @@ private boolean exploreNode(int group, Context context, Matcher matcher)
                     done = false;
                     progress = true;
                 }
+
+                if (result.getTrait().isPresent()) {
+                    Trait trait = result.getTrait().get();
+
+                    checkState(trait.getType().isEquivalenceGroupApplicable(), "Only group trait are currently supported");
+                    context.getMemo().setGroupTrait(group, trait);
```

Comment:
I think for starters we could just call `addGroupTrait` and verify that the existing (if present) trait of the same type is equal. That is how I understand `isEquivalenceGroupApplicable`. Trait can be propagated to entire group and doesn't change, which means the optimizer can skip computing the trait of a given type.